daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala

```diff
@@ -112,83 +148,61 @@ abstract class InfosetInputter
    */
   val supportsNamespaces: Boolean
 
+  final override lazy val advanceAccessor = InfosetAccessor()
+  final override lazy val inspectAccessor = InfosetAccessor()
+
+  private var documentElement_ : DIDocument = _
+
+  /**
+   * Mostly this is for unit testing where we want to inspect
+   * the infoset the unparser constructed, as well as looking at
+   * the data that was output.
+   */
+  def documentElement = {
+    Assert.usage(isInitialized, "Must be initialized")
+    documentElement_
+  }
+
+  /**
+   * Initialize the infoset inputter, given the ERD of the root element, and
+   * the tunables.
+   */
   def initialize(rootElementInfo: ElementRuntimeData, tunableArg: DaffodilTunables): Unit = {
     tunable = tunableArg
-    nextElementResolver =
-      new OnlyOnePossibilityForNextElement(rootElementInfo.schemaFileLocation, rootElementInfo, RootResolver) // bootstrap
+    isInitialized_ = true
 
-    val diDoc = new DIDocument(rootElementInfo, tunable)
-    nodeStack.push(diDoc)
+    documentElement_ = new DIDocument(rootElementInfo, tunable)
+    infoStack.push(documentElement_)
 
     try {
-      if (!hasNext() || getEventType() != StartDocument) {
-        UnparseError(One(nodeStack.top.erd.schemaFileLocation), Nope, "Infoset does not start with StartDocument event")
+      if (!hasNext() || getEventType() != InfosetInputterEventType.StartDocument) {
+        UnparseError(One(infoStack.top.erd.schemaFileLocation), Nope, "Infoset does not start with StartDocument event")
       }
     } catch {
       case e: InvalidInfosetException =>
-        UnparseError(One(nodeStack.top.erd.schemaFileLocation), Nope, "Infoset does not start with StartDocument event: " + e.getMessage)
+        UnparseError(One(infoStack.top.erd.schemaFileLocation), Nope, "Infoset does not start with StartDocument event: " + e.getMessage)
     }
-  }
 
-  private type NodeStack = MStackOf[DINode]
-  private val nodeStack: NodeStack = new MStackOf[DINode]
+    this.pushTRD(rootElementInfo) // TRD stack holds TRDs of model groups
+  }
 
-  private var nextElementResolver: NextElementResolver = null
+  private val infoStack = new Info.InfoStack
 
-  private def level = nodeStack.length
+  private def level = infoStack.length
 
   private def indent = ("  " * level) + "|"
 
   override def toString() = {
     indent + "************* STATE ************************\n" +
-      indent + "nextElementResolver = " + nextElementResolver + "\n" +
-      indent + "nodeStack = " + nodeStack + "\n" +
+      indent + "infoStack = " + infoStack + "\n" +
       indent + "********************************************"
   }
 
-  override def inspectPure: Boolean = {
```

Comment:
There is still an InspectPure case in ./daffodil-lib/src/main/scala/org/apache/daffodil/util/Cursor.scala. I assume we want to remove that as well?

Also, check comments in ./daffodil-lib/src/main/scala/org/apache/daffodil/util/Cursor.scala and ./daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala for lingering references.

Reply:
Thanks. I removed all the inspectPure references. Next commit.